### PR TITLE
applepay updated

### PR DESCRIPTION
**Pull request by andyfen**
based off the design made by @sbkexperiments

here 
https://hackmd.io/4Up6M-BIR1WTzUirtuBF4A?view

#### Integration 
```
function onClick() {
  const applePayPaymentRequest = {
    countryCode: window.applePayConfig.countryCode,
    currencyCode: window.applePayConfig.currencyCode,
    merchantCapabilities: window.applePayConfig.merchantCapabilities,
    supportedNetworks: window.applePayConfig.supportedNetworks,
    requiredShippingContactFields: ["name", "phone", "email", "postalAddress"],
    total: {
      label: "Demo",
      type: "final",
      amount: "99.99",
    },
  };

  var session = new ApplePaySession(4, applePayPaymentRequest);

  session.onvalidatemerchant = (event) => {
    paypal.applePay
      .validateMerchant({
        validationUrl: event.validationURL,
      })
      .then((merchantSession) => {
        const session = atob(merchantSession.session);
        session.completeMerchantValidation(merchantSession);
      })
      .catch(() => {
        session.abort();
      });
  };

  session.onshippingcontactselected = (event) => {
    const shippingContactUpdate = {};
    session.completeShippingContactSelection(shippingContactUpdate);
  };

  session.onshippingmethodselected = (event) => {
    console.log("Your shipping method is:", event.shippingMethod);
    // Update payment details.
    var shippingMethodUpdate = {}; // https://developer.apple.com/documentation/apple_pay_on_the_web/applepayshippingmethodupdate
    session.completeShippingMethodSelection(shippingMethodUpdate); // Set shippingMethodUpdate=null if there are no updates.
  };

  session.onpaymentauthorized = async (event) => {
    try {
      const res = await fetch("https://your-server.com/orders", {
        method: "POST",
        body: JSON.stringify({
          purchase_units: [
            {
              amount: {
                currency_code: "USD",
                value: "99.99",
              },
            },
          ],
        }),
      })
      if(!res.ok){
        throw new Error("order create failed")
      }

      const { id: orderId } = await res.json()

      await paypal.applePay.confirmOrder({
        orderId,
        token: event.payment.token,
        billingContact: event.payment.billingContact, // or provide billing address collected on merchant page.
        shippingContact: event.payment.shippingContact,
      });

      session.completePayment(ApplePaySession.STATUS_SUCCESS);

      //Submit approval to the server and authorize or capture the order.
      await fetch(`https://your-server.com/capture/${orderId}`, {
        method: "post",
      });
    } catch (err) {
      session.completePayment(ApplePaySession.STATUS_FAILURE);
    }
  };

  // Open Apple sheet.
  session.begin();
}

if (
  window.ApplePaySession &&
  ApplePaySession.supportsVersion(4) &&
  ApplePaySession.canMakePayments()
) {
  (async () => {
    try {
      window.applePayConfig = await paypal.applePay.getConfiguration();

      if (!window.applePayConfig.isApplePayEligible) {
        return;
      }

      const applePayButton = document.getElementById("applepay-btn");
      applePayButton.innerHTML =
        '<apple-pay-button buttonstyle="black" type="buy" locale="el-GR">';

      applePayButton.addEventListener("click", onClick);
    } catch (err) {
      console.error(err);
    }
  })();
}

```

### CODE REVIEW

**Comment by andyfen** [comment]
couple of thought 

- the order details get passed in twice in 2 separate formats `apple` and `paypal` into applepayPaymentRequest and order create ?

alternative initialization idea
```
paypal.Applepay({ clientId: "1", merchantId: "2" })
```

**Review by mnicpt — CHANGES_REQUESTED** [review]
@andyfen This is how any merchant would Apple Pay to their site.  Why use the `paypal` global object to just provide a config and create an order.  Can't the merchant do that already without a helper method on our global object?  The merchant is already doing server-side integration.  Have them create the order server-side as well. Definitely love where we want to go, and want to make sure we have all the stakeholders involved to move forward in the right way.  This is a big need to have a headless integration.

**Review by mnicpt — APPROVED** [review]

File: spec/client/javascript/applepay/advanced.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+if (
+  window.ApplePaySession &&
+  ApplePaySession.supportsVersion(4) &&
+  ApplePaySession.canMakePayments()
+) {
+  (async () => {
+    try {
+      window.applePayConfig = await paypal.applePay.getConfiguration();
```
Comment (mnicpt) [inline spec/client/javascript/applepay/advanced.md:99]:
```suggestion
      const applePayConfig = await paypal.applePay.getConfiguration();
```

File: spec/client/javascript/applepay/advanced.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        orderId,
+        token: event.payment.token,
+        billingContact: event.payment.billingContact, // or provide billing address collected on merchant page.
+        shippingContact: event.payment.shippingContact,
+      });
+
+      session.completePayment(ApplePaySession.STATUS_SUCCESS);
+
+      //Submit approval to the server and authorize or capture the order.
+      await fetch(`https://your-server.com/capture/${orderId}`, {
```
Comment (mnicpt) [inline spec/client/javascript/applepay/advanced.md:80]: @andyfen Why are we providing convenience methods for some calls but not all?  I see them for `validateMerchant`, `confirmPayment` but not for createOrder or capture call.

Comment (mnicpt) [reply spec/client/javascript/applepay/advanced.md:80]: Love `createOrder` here!!!

**Comment by andyfen** [comment]
if this was to be loaded via a "component" 

that name scope would look different it would be `paypal.Applepay` not `paypal.applepay.methodName`

it may need to look more like 
```
const applepay = paypal.Applepay({
    clientId: "xxx",
    merchantId: "xxx",
})

const config = await applepay.getConfiguration({})

session.onvalidatemerchant = async (event) => {
   const merchantSession =  await applepay.validateMerchant()
   session.completeMerchantValidation(merchantSession);
}

session.onpaymentauthorized = async (event) => {
    const { id } = await fetch("/api/orders", {
        method: "POST",
        // ....
    }).then(res => res.json())

    await applepay.confirmOrder({
        orderId: orderId,
        token: event.payment.token,
        billingContact: event.payment.billingContact,
    })
    session.completePayment(ApplePaySession.STATUS_SUCCESS);
}

session.onshippingmethodselected = async (event) => {
    // ....
}
```

**Comment by sbkexperiments** [comment]
> const applepay = paypal.Applepay({
>     clientId: "xxx",
>     merchantId: "xxx",
> })

LGTM.

Like the 
```
const applepay = paypal.Applepay({
    clientId: "xxx",
    merchantId: "xxx",
})
```

**Comment by andyfen** [comment]
just an fyi `createOrder` can be called on the client with a low scoped access token, just the base64 client id (no secret), so it could have this option also

```
const { id } = await applepay.createOrder({
  purchase_units: [
    {
      amount: {
        currency_code: "USD",
        value: "0.01",
      },
    },
  ],
})
```

Access token
```
  const basicAuth = btoa(`${clientId}:`); // No secret required

  return fetch("/v1/oauth2/token", {
    method: "POST",
    headers: {
      Authorization: `Basic ${basicAuth}`,
    },
    agent: httpsAgent,
    body: "grant_type=client_credentials",
  })
    .then((res) => res.json())
    .then((res) => {
      return res.access_token;
    });
```

**Comment by sioked** [comment]
Still a bit of an open question on if this is loaded as a component or if it's loaded as a util library. `paypal.applePay` vs `paypal.ApplePay()`.

I like the pattern of:
```
const applepay = paypal.Applepay({
    clientId: "xxx",
    merchantId: "xxx",
})
```

But that is inconsistent with the way that we initialize the SDK in other places today. I also think that in most cases the existing components that are capitalized are UI components, ie `paypal.Buttons()` or `paypal.Messages()`. I think for util libraries it might make sense to use existing pattern and to use lowercase namespaces to imply that it's a collection of util methods. Specifically `paypal.applePay.foo`.

**Review by sioked — APPROVED** [review]

File: spec/client/javascript/applepay/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Applepay
+
+- [Advaanced](./advanced.md)
```
Comment (gregjopa) [inline spec/client/javascript/applepay/README.md:3]:
```suggestion
- [Advanced](./advanced.md)
```

File: spec/client/javascript/README.md [hunk 4]
```diff
@@ -7,3 +7,6 @@
 - [Components](./components/README.md)
 - [Callbacks](./callbacks/README.md)
 - [Constants](./constants/README.md)
+
+
+- [Applepaay](./applepay/README.md)
```
Comment (gregjopa) [inline spec/client/javascript/README.md:12]:
```suggestion
- [Applepay](./applepay/README.md)
```

File: spec/client/javascript/applepay/advanced.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
+# Applepay Advanced 
+
```
Comment (gregjopa) [inline spec/client/javascript/applepay/advanced.md:2]: Can we add a section to show how the JS SDK script will be loaded for ApplePay? The main thing to call out is the new proposed "components=applepayweb" component

```html
<script src="https://www.paypal.com/sdk/js?client-id=YOUR_CLIENT_ID&currency=USD&buyer-country=US&merchant-id=SUB_MERCHANT_ID&components=applepayweb"></script>
```

File: spec/client/javascript/applepay/advanced.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  (async () => {
+    try {
+      window.applePayConfig = await paypal.applePay.getConfiguration();
+
+      if (!window.applePayConfig.isApplePayEligible) {
+        return;
+      }
+
+      const applePayButton = document.getElementById("applepay-btn");
+      applePayButton.innerHTML =
```
Comment (gregjopa) [inline spec/client/javascript/applepay/advanced.md:106]: Can we add comment here explaining that this ApplePay button will be rendered directly into the merchant's web page? I want to make sure the reader understands that this proposed design will not be wrapped inside an iframe-hosted on paypal.com.

**Review by gregjopa — APPROVED** [review]

**Pull request merged**